File: Core/Install/IPAParser.m
#import "IPAParser.h"
#import <zlib.h>
#import "ChoMAWrapper.h"
#import "Logger.h"

typedef struct __attribute__((packed)) {
    uint32_t signature;
    uint16_t versionNeeded;
    uint16_t flags;
    uint16_t compression;
    uint16_t modTime;
    uint16_t modDate;
    uint32_t crc32;
    uint32_t compressedSize;
    uint32_t uncompressedSize;
    uint16_t fileNameLen;
    uint16_t extraFieldLen;
} ZIPLocalHeader;

typedef struct __attribute__((packed)) {
    uint32_t signature;
    uint16_t versionMade;
    uint16_t versionNeeded;
    uint16_t flags;
    uint16_t compression;
    uint16_t modTime;
    uint16_t modDate;
    uint32_t crc32;
    uint32_t compressedSize;
    uint32_t uncompressedSize;
    uint16_t fileNameLen;
    uint16_t extraFieldLen;
    uint16_t commentLen;
    uint16_t diskStart;
    uint16_t internalAttrs;
    uint32_t externalAttrs;
    uint32_t localOffset;
} ZIPCentralDir;

static const NSUInteger CHUNK_SIZE = 4 * 1024 * 1024; // 4MB streaming chunks

@implementation IPAParser

+ (NSData *)extractCDHashFromIPAAt:(NSURL *)ipaPath error:(NSError **)error {
    NSURL *tmpDir = [NSURL fileURLWithPath:NSTemporaryDirectory()];
    tmpDir = [tmpDir URLByAppendingPathComponent:[[NSUUID UUID] UUIDString]];
    [[NSFileManager defaultManager] createDirectoryAtURL:tmpDir withIntermediateDirectories:YES attributes:nil error:nil];

    if (![self unzipIPAAt:ipaPath to:tmpDir error:error]) {
        [[NSFileManager defaultManager] removeItemAtURL:tmpDir error:nil];
        return nil;
    }

    NSURL *payloadDir = [tmpDir URLByAppendingPathComponent:@"Payload"];
    NSArray *contents = [[NSFileManager defaultManager] contentsOfDirectoryAtPath:payloadDir.path error:nil];
    NSString *appDir = nil;
    for (NSString *name in contents) {
        if ([name hasSuffix:@".app"]) { appDir = name; break; }
    }
    if (!appDir) {
        if (error) *error = [NSError errorWithDomain:@"IPA" code:-1 userInfo:@{NSLocalizedDescriptionKey: @"Missing .app in Payload"}];
        [[NSFileManager defaultManager] removeItemAtURL:tmpDir error:nil];
        return nil;
    }

    NSURL *appPath = [payloadDir URLByAppendingPathComponent:appDir];
    NSURL *infoPlistURL = [appPath URLByAppendingPathComponent:@"Info.plist"];
    NSData *infoData = [NSData dataWithContentsOfURL:infoPlistURL];
    if (!infoData) {
        if (error) *error = [NSError errorWithDomain:@"IPA" code:-1 userInfo:@{NSLocalizedDescriptionKey: @"Missing Info.plist"}];
        [[NSFileManager defaultManager] removeItemAtURL:tmpDir error:nil];
        return nil;
    }
    NSDictionary *info = [NSPropertyListSerialization propertyListWithData:infoData options:0 format:nil error:error];
    if (!info) {
        [[NSFileManager defaultManager] removeItemAtURL:tmpDir error:nil];
        return nil;
    }
    NSString *execName = info[@"CFBundleExecutable"];
    if (!execName) {
        if (error) *error = [NSError errorWithDomain:@"IPA" code:-1 userInfo:@{NSLocalizedDescriptionKey: @"No CFBundleExecutable"}];
        [[NSFileManager defaultManager] removeItemAtURL:tmpDir error:nil];
        return nil;
    }

    NSString *execPath = [[appPath URLByAppendingPathComponent:execName] path];
    NSData *cdhash = [ChoMAWrapper extractCDHash:execPath];
    [[NSFileManager defaultManager] removeItemAtURL:tmpDir error:nil];

    if (!cdhash) {
        if (error) *error = [NSError errorWithDomain:@"IPA" code:-1 userInfo:@{NSLocalizedDescriptionKey: @"CDHash extraction failed"}];
    }
    return cdhash;
}

+ (BOOL)unzipIPAAt:(NSURL *)source to:(NSURL *)dest error:(NSError **)error {
    NSFileHandle *fh = [NSFileHandle fileHandleForReadingFromURL:source error:error];
    if (!fh) return NO;

    // Read EOCD from tail — stream-safe: only need last 64KB
    uint64_t fileSize = [fh seekToEndOfFile];
    if (fileSize < 22) {
        if (error) *error = [NSError errorWithDomain:@"IPA" code:-1 userInfo:@{NSLocalizedDescriptionKey: @"File too small"}];
        [fh closeFile];
        return NO;
    }

    NSUInteger searchLen = (NSUInteger)MIN(fileSize, 65557);
    uint64_t searchStart = fileSize - searchLen;
    [fh seekToFileOffset:searchStart];
    NSData *tail = [fh readDataOfLength:searchLen];
    if (tail.length != searchLen) {
        if (error) *error = [NSError errorWithDomain:@"IPA" code:-1 userInfo:@{NSLocalizedDescriptionKey: @"Cannot read EOCD region"}];
        [fh closeFile];
        return NO;
    }

    NSInteger eocdOffset = -1;
    for (NSInteger i = 0; i < (NSInteger)searchLen - 3; i++) {
        uint32_t sig;
        [tail getBytes:&sig range:NSMakeRange(i, 4)];
        if (sig == 0x06054b50) { eocdOffset = (NSInteger)searchStart + i; break; }
    }
    if (eocdOffset < 0) {
        if (error) *error = [NSError errorWithDomain:@"IPA" code:-1 userInfo:@{NSLocalizedDescriptionKey: @"Invalid ZIP: no EOCD"}];
        [fh closeFile];
        return NO;
    }

    // Read central directory from EOCD
    uint16_t numEntries;
    uint32_t cdOffset;
    [tail getBytes:&numEntries range:NSMakeRange((NSUInteger)(eocdOffset - (NSInteger)searchStart) + 10, 2)];
    [tail getBytes:&cdOffset range:NSMakeRange((NSUInteger)(eocdOffset - (NSInteger)searchStart) + 16, 4)];

    // Read central directory into memory (typically < 64KB)
    [fh seekToFileOffset:cdOffset];
    NSUInteger cdBytes = (NSUInteger)(eocdOffset - (NSInteger)cdOffset);
    NSData *centralDir = [fh readDataOfLength:cdBytes];
    if (centralDir.length != cdBytes) {
        [fh closeFile];
        if (error) *error = [NSError errorWithDomain:@"IPA" code:-1 userInfo:@{NSLocalizedDescriptionKey: @"Cannot read central directory"}];
        return NO;
    }

    NSInteger pos = 0;
    for (uint16_t i = 0; i < numEntries; i++) {
        if ((NSUInteger)pos + sizeof(ZIPCentralDir) > centralDir.length) break;
        ZIPCentralDir cd;
        [centralDir getBytes:&cd range:NSMakeRange(pos, sizeof(ZIPCentralDir))];
        if (cd.signature != 0x02014b50) break;

        NSString *name = [[NSString alloc] initWithData:[centralDir subdataWithRange:NSMakeRange(pos + 46, cd.fileNameLen)] encoding:NSUTF8StringEncoding];
        if (!name) { pos += 46 + cd.fileNameLen + cd.extraFieldLen + cd.commentLen; continue; }
        pos += 46 + cd.fileNameLen + cd.extraFieldLen + cd.commentLen;

        // Zip slip guard
        if ([name containsString:@".."] || [name hasPrefix:@"/"]) continue;

        NSURL *destPath = [dest URLByAppendingPathComponent:name];
        if ([name hasSuffix:@"/"]) {
            [[NSFileManager defaultManager] createDirectoryAtURL:destPath withIntermediateDirectories:YES attributes:nil error:nil];
            continue;
        }
        [[NSFileManager defaultManager] createDirectoryAtURL:[destPath URLByDeletingLastPathComponent] withIntermediateDirectories:YES attributes:nil error:nil];

        // Read local header via stream
        [fh seekToFileOffset:cd.localOffset];
        ZIPLocalHeader lh;
        NSData *lhData = [fh readDataOfLength:sizeof(ZIPLocalHeader)];
        if (lhData.length != sizeof(ZIPLocalHeader)) continue;
        [lhData getBytes:&lh length:sizeof(ZIPLocalHeader)];
        if (lh.signature != 0x04034b50) continue;

        uint64_t dataOff = cd.localOffset + 30 + lh.fileNameLen + lh.extraFieldLen;
        [fh seekToFileOffset:dataOff];

        // Stream file data to disk — never hold entire compressed entry in RAM
        if (cd.compression == 0) {
            [[NSFileManager defaultManager] createFileAtPath:destPath.path contents:nil attributes:nil];
            NSFileHandle *outFH = [NSFileHandle fileHandleForWritingAtPath:destPath.path];
            if (!outFH) continue;

            uint64_t remaining = cd.compressedSize;
            while (remaining > 0) {
                NSUInteger toRead = (NSUInteger)MIN(remaining, CHUNK_SIZE);
                NSData *chunk = [fh readDataOfLength:toRead];
                if (chunk.length == 0) break;
                [outFH writeData:chunk];
                remaining -= chunk.length;
            }
            [outFH closeFile];
        } else if (cd.compression == 8) {
            // For deflate, buffer entire entry + decompress to file
            NSMutableData *compBuf = [NSMutableData dataWithCapacity:(NSUInteger)cd.compressedSize];
            uint64_t remaining = cd.compressedSize;
            while (remaining > 0) {
                NSUInteger toRead = (NSUInteger)MIN(remaining, CHUNK_SIZE);
                NSData *chunk = [fh readDataOfLength:toRead];
                if (chunk.length == 0) break;
                [compBuf appendData:chunk];
                remaining -= chunk.length;
            }

            if (compBuf.length == cd.compressedSize) {
                NSData *decompressed = [self decompressDeflate:compBuf uncompressedSize:cd.uncompressedSize];
                if (decompressed) [decompressed writeToURL:destPath atomically:YES];
            }
        }
    }

    [fh closeFile];
    return YES;
}

+ (NSData *)decompressDeflate:(NSData *)compressed uncompressedSize:(uint32_t)uncompressedSize {
    if (uncompressedSize == 0) return nil;
    NSMutableData *result = [NSMutableData dataWithLength:uncompressedSize];
    z_stream stream;
    memset(&stream, 0, sizeof(stream));
    int ret = inflateInit2_(&stream, -15, ZLIB_VERSION, sizeof(stream));
    if (ret != Z_OK) return nil;

    stream.next_in = (Bytef *)compressed.bytes;
    stream.avail_in = (uInt)compressed.length;
    stream.next_out = (Bytef *)result.mutableBytes;
    stream.avail_out = (uInt)uncompressedSize;
    ret = inflate(&stream, Z_FINISH);
    inflateEnd(&stream);

    return (ret == Z_STREAM_END) ? result : nil;
}

@end

File: Core/Install/IPAParser.swift
import Foundation
import zlib

class IPAParser {
    static func extractCDHash(ipaPath: URL) -> Data? {
        // 1. Giải nén IPA vào tmp
        let tmpDir = FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString)
        defer { try? FileManager.default.removeItem(at: tmpDir) }
        guard (try? unzipIPA(at: ipaPath, to: tmpDir)) != nil else { return nil }

        // Tìm file .app bên trong Payload/
        let payloadDir = tmpDir.appendingPathComponent("Payload")
        guard let contents = try? FileManager.default.contentsOfDirectory(atPath: payloadDir.path) else { return nil }
        guard let appDir = contents.first(where: { $0.hasSuffix(".app") }) else { return nil }
        let appPath = payloadDir.appendingPathComponent(appDir)

        // Đọc Info.plist để lấy CFBundleExecutable
        let infoPlist = appPath.appendingPathComponent("Info.plist")
        guard let infoData = try? Data(contentsOf: infoPlist) else { return nil }
        guard let info = try? PropertyListSerialization.propertyList(from: infoData, format: nil) as? [String: Any] else { return nil }
        guard let execName = info["CFBundleExecutable"] as? String else { return nil }
        let execPath = appPath.appendingPathComponent(execName)

        // 2. Lấy CDHash của file binary chính
        return ChoMAWrapper.extractCDHash(execPath.path)
    }

    static func unzipIPA(at source: URL, to dest: URL) throws {
        let data = try Data(contentsOf: source)
        let searchLen = min(data.count, 65557)
        let start = data.count - searchLen
        var eocdOffset = -1
        for i in 0..<searchLen - 3 {
            if try data.readUInt32(at: start + i) == 0x06054b50 {
                eocdOffset = start + i
                break
            }
        }
        guard eocdOffset >= 0 else { throw NSError(domain: "IPA", code: -1) }

        let numEntries = Int(try data.readUInt16(at: eocdOffset + 10))
        let cdOffset = Int(try data.readUInt32(at: eocdOffset + 16))

        var pos = cdOffset
        for _ in 0..<numEntries {
            guard try data.readUInt32(at: pos) == 0x02014b50 else { break }
            let compression = try data.readUInt16(at: pos + 10)
            let compSize = Int(try data.readUInt32(at: pos + 20))
            let uncompSize = Int(try data.readUInt32(at: pos + 24))
            let nameLen = Int(try data.readUInt16(at: pos + 28))
            let extraLen = Int(try data.readUInt16(at: pos + 30))
            let localOff = Int(try data.readUInt32(at: pos + 42))
            let name = try data.readString(at: pos + 46, length: nameLen)
            pos += 46 + nameLen + extraLen + Int(try data.readUInt16(at: pos + 32))

            guard !name.contains("..") && !name.hasPrefix("/") else { continue }
            let destPath = dest.appendingPathComponent(name)
            if name.hasSuffix("/") {
                try FileManager.default.createDirectory(at: destPath, withIntermediateDirectories: true)
                continue
            }
            try FileManager.default.createDirectory(at: destPath.deletingLastPathComponent(), withIntermediateDirectories: true)

            guard try data.readUInt32(at: localOff) == 0x04034b50 else { continue }
            let lNameLen = Int(try data.readUInt16(at: localOff + 26))
            let lExtraLen = Int(try data.readUInt16(at: localOff + 28))
            let dataOff = localOff + 30 + lNameLen + lExtraLen
            let rawData = data[dataOff..<dataOff + compSize]

            if compression == 0 {
                try rawData.write(to: destPath)
            } else if compression == 8, let decompressed = decompressDeflate(rawData, uncompressedSize: uncompSize) {
                try decompressed.write(to: destPath)
            }
        }
    }

    private static func decompressDeflate(_ compressed: Data, uncompressedSize: Int) -> Data? {
        guard uncompressedSize > 0 else { return nil }
        var result = Data(count: uncompressedSize)
        var stream = z_stream()
        let initStatus = inflateInit2_(&stream, -15, ZLIB_VERSION, Int32(MemoryLayout<z_stream>.size))
        guard initStatus == Z_OK else { return nil }
        defer { inflateEnd(&stream) }

        let ret = compressed.withUnsafeBytes { srcBuf in
            result.withUnsafeMutableBytes { dstBuf in
                guard let srcBase = srcBuf.baseAddress?.assumingMemoryBound(to: UInt8.self),
                      let dstBase = dstBuf.baseAddress?.assumingMemoryBound(to: UInt8.self) else {
                    return Z_STREAM_ERROR
                }
                stream.next_in = UnsafeMutablePointer(mutating: srcBase)
                stream.avail_in = uInt(compressed.count)
                stream.next_out = dstBase
                stream.avail_out = uInt(uncompressedSize)
                return inflate(&stream, Z_FINISH)
            }
        }
        return ret == Z_STREAM_END ? result : nil
    }
}

extension Data {
    func readUInt32(at offset: Int) throws -> UInt32 {
        guard offset + 4 <= count else { throw NSError(domain: "read", code: -1) }
        return UInt32(self[offset]) | (UInt32(self[offset+1]) << 8) | (UInt32(self[offset+2]) << 16) | (UInt32(self[offset+3]) << 24)
    }

    func readUInt16(at offset: Int) throws -> UInt16 {
        guard offset + 2 <= count else { throw NSError(domain: "read", code: -1) }
        return UInt16(self[offset]) | (UInt16(self[offset+1]) << 8)
    }

    func readString(at offset: Int, length: Int) throws -> String {
        guard offset + length <= count else { throw NSError(domain: "read", code: -1) }
        return String(data: self[offset..<offset+length], encoding: .utf8) ?? ""
    }
}
